frame/contracts/src/lib.rs

```diff
@@ -385,6 +392,34 @@ pub mod pallet {
 			output.gas_meter.into_dispatch_result(output.result, T::WeightInfo::call())
 		}
 
+		#[pallet::weight(T::WeightInfo::call().saturating_add(*gas_limit))]
+		pub fn query(
+			origin: OriginFor<T>,
+			dest: <T::Lookup as StaticLookup>::Source,
+			#[pallet::compact] value: BalanceOf<T>,
+			#[pallet::compact] gas_limit: Weight,
+			storage_deposit_limit: Option<<BalanceOf<T> as codec::HasCompact>::Type>,
+			data: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let origin = ensure_signed(origin)?;
+			let dest = T::Lookup::lookup(dest)?;
+			let mut output = Self::internal_query(
+				origin,
+				dest,
+				value,
+				gas_limit,
+				storage_deposit_limit.map(Into::into),
+				data,
+				None,
+			);
+			if let Ok(retval) = &output.result {
+				if retval.did_revert() {
+					output.result = Err(<Error<T>>::ContractReverted.into());
+				}
+			}
+			output.gas_meter.into_dispatch_result(output.result, T::WeightInfo::call())
+		}
```

Comment:
I think query can be an rpc method; or does internal_query allow for state changes?

Reply:
It just a copy/paste of execute right now